go/ephemeral/device_ek_storage.go

```diff
@@ -0,0 +1,318 @@
+	boxed := boxedData{
+		V: cryptoVersion,
+		E: sealed,
+		N: fnonce,
+	}
+
+	// Encode encrypted box
+	return encode(boxed)
+}
+
+func (s *DeviceEKStorage) write(generation keybase1.EkGeneration, data []byte) (err error) {
+	filepath := s.filepath(generation)
+
+	if err := os.MkdirAll(s.storagePath, libkb.PermDir); err != nil {
+		return err
+	}
+	tmp, err := ioutil.TempFile(s.storagePath, fmt.Sprintf("%s-%d", deviceEKPrefix, generation))
+	if err != nil {
+		return err
+	}
+	// remove the temp file if it still exists at the end of this function
+	defer libkb.ShredFile(tmp.Name())
+
+	if runtime.GOOS != "windows" {
+		// os.Fchmod not supported on windows
+		if err := tmp.Chmod(libkb.PermFile); err != nil {
+			return err
+		}
+	}
+	if _, err := tmp.Write(data); err != nil {
+		return err
+	}
+	if err := tmp.Close(); err != nil {
+		return err
+	}
+
+	// remove the secret if it already exsists
+	s.delete(generation)
+
+	if err := os.Rename(tmp.Name(), filepath); err != nil {
+		return err
+	}
+
+	if runtime.GOOS != "windows" {
+		// os.Fchmod not supported on windows
+		if err := os.Chmod(filepath, libkb.PermFile); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *DeviceEKStorage) Delete(generation keybase1.EkGeneration) error {
+	s.Lock()
+	defer s.Unlock()
+	return s.delete(generation)
+}
+
+func (s *DeviceEKStorage) delete(generation keybase1.EkGeneration) error {
+	// clear the cache
+	delete(s.keys, generation)
+
+	filepath := s.filepath(generation)
+	exists, err := libkb.FileExists(filepath)
+	if err != nil {
+		return err
+	}
+	if exists {
+		err = libkb.ShredFile(filepath)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *DeviceEKStorage) index(ctx context.Context) (err error) {
+	s.indexOnce.Do(func() {
+		files, err := filepath.Glob(filepath.Join(s.storagePath, "*.ek"))
+		if err != nil {
+			return
+		}
+
+		for _, file := range files {
+			if strings.HasPrefix(file, deviceEKPrefix) {
+				parts := strings.Split(file, deviceEKPrefix)
+				g, err := strconv.ParseUint(parts[1], 10, 64)
+				if err != nil {
+					return
+				}
+				generation := keybase1.EkGeneration(g)
+				deviceEK, err := s.Get(ctx, generation)
+				if err != nil {
+					return
+				}
+				s.keys[generation] = deviceEK
+			}
+		}
+	})
+	return err
+}
+
+func (s *DeviceEKStorage) GetAll(ctx context.Context) (deviceEKs map[keybase1.EkGeneration]keybase1.DeviceEk, err error) {
+	s.Lock()
+	defer s.Unlock()
+
+	err = s.index(ctx)
+	return s.keys, err
+}
+
+func (s *DeviceEKStorage) MaxGeneration(ctx context.Context) (maxGeneration keybase1.EkGeneration) {
+	s.Lock()
+	defer s.Unlock()
+
+	err := s.index(ctx)
+	if err != nil {
+		return maxGeneration
+	}
+	for generation := range s.keys {
+		if generation > maxGeneration {
+			maxGeneration = generation
+		}
+	}
+	return maxGeneration
+}
+
+func getLocalStorageSecretBoxKey(ctx context.Context, g *libkb.GlobalContext) (fkey [32]byte, err error) {
```

Comment:
so the device key files themselves are encrypted via lksec.  You've probably discussed amongst yourselves, but what is the reason not to use that for these files as well?

Reply:
Since the device long term keys are LKSec'd, I think encrypting something with the long term keys is roughly equivalent to LKSec'ing it? So I just went with that cause it felt simple. Do you have a preference one way or the other? (Chat storage also uses the device long term key, right?)